server/keyspace/keyspace.go

```diff
@@ -0,0 +1,244 @@
+	// Validate input.
+	if request.Name == "" {
+		return nil, errIllegalName
+	}
+	// Allocate new keyspaceID.
+	newID, err := manager.allocID()
+	if err != nil {
+		return nil, err
+	}
+	// Create and save keyspace metadata.
+	keyspace := &keyspacepb.KeyspaceMeta{
+		Id:             newID,
+		Name:           request.Name,
+		State:          keyspacepb.KeyspaceState_ENABLED,
+		CreatedAt:      request.Now.Unix(),
+		StateChangedAt: request.Now.Unix(),
+		Config:         request.InitialConfig,
+	}
+	manager.metaLock.Lock()
+	defer manager.metaLock.Unlock()
+	if err := manager.store.SaveKeyspace(keyspace); err != nil {
+		return nil, err
+	}
+	// Create name to ID entry,
+	// if this failed, previously stored keyspace meta should be removed.
+	if err = manager.createNameToID(newID, request.Name); err != nil {
+		if removeErr := manager.store.RemoveKeyspace(newID); removeErr != nil {
+			return nil, errors.Wrap(removeErr, "failed to remove keyspace meta after save spaceID failure")
+		}
+		return nil, err
+	}
+
+	return keyspace, nil
+}
+
+// loadKeyspaceID returns the id of keyspace specified by name.
+// It returns error if loading met error, or if keyspace not exists.
+func (manager *Manager) loadKeyspaceID(name string) (uint32, error) {
+	loaded, spaceID, err := manager.store.LoadKeyspaceID(name)
+	if err != nil {
+		return 0, err
+	}
+	if !loaded {
+		return 0, ErrKeyspaceNotFound
+	}
+	return spaceID, nil
+}
+
+// LoadKeyspace returns the keyspace specified by name.
+// It returns error if loading or unmarshalling met error or if keyspace does not exist.
+func (manager *Manager) LoadKeyspace(name string) (*keyspacepb.KeyspaceMeta, error) {
+	spaceID, err := manager.loadKeyspaceID(name)
+	if err != nil {
+		return nil, err
+	}
+	keyspace := &keyspacepb.KeyspaceMeta{}
+	loaded, err := manager.store.LoadKeyspace(spaceID, keyspace)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	return keyspace, nil
+}
+
+// UpdateKeyspace update keyspace's state and config according to request.
+// It returns error if saving failed, operation not allowed, or if keyspace not exists.
+func (manager *Manager) UpdateKeyspace(request *UpdateKeyspaceRequest) (*keyspacepb.KeyspaceMeta, error) {
+	manager.metaLock.Lock()
+	defer manager.metaLock.Unlock()
+	// load keyspace by id
+	keyspace, err := manager.LoadKeyspace(request.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// disallow modifying archived keyspace
+	if keyspace.State == keyspacepb.KeyspaceState_ARCHIVED {
+		return nil, errKeyspaceArchived
+	}
+
+	// update keyspace state and config
+	if request.UpdateState {
+		if err = changeKeyspaceState(keyspace, request.NewState, request.Now); err != nil {
+			return nil, err
+		}
+	}
+	changeKeyspaceConfig(keyspace, request.ToPut, request.ToDelete)
+	// save keyspace
+	if err = manager.store.SaveKeyspace(keyspace); err != nil {
```

Comment:
If the change of state and config is already saved in front. 
Is there any change for KeySpace ?

Reply:
Is it need complete keyspace information in request?